cli/src/commands/generate/schema-directory.ts

```diff
@@ -0,0 +1,142 @@
+    private readonly schemas: Map<string, any> = new Map<string, any>();
+    private readonly logger: Logger;
+
+    /**
+     * Initialise the SchemaDirectory. Does not load the schemas until loadSchemas is called.
+     * @param directoryPath The directory path from which to load schemas. All JSON and YAML files under this path will be loaded, including subfolders.
+     * @param debug Whether to log at debug level.
+     */
+    constructor(private directoryPath: string, debug: boolean = false) {
+        this.logger = initLogger(debug);
+    }
+
+    /**
+     * Load the schemas from the configured directory path.
+     */
+    public async loadSchemas() {
+        this.logger.debug('Loading schemas from ' + this.directoryPath);
+        const files = await readdir(this.directoryPath, { recursive: true });
+
+        const schemaPaths = files.filter(str => str.match(/^.*(json|yaml|yml)$/))
+            .map(schemaPath => join(this.directoryPath, schemaPath));
+
+        for (const schemaPath of schemaPaths) {
+            await this.loadSchema(schemaPath);
+        }
+
+        this.logger.info(`Loaded ${this.schemas.size} schemas.`);
+    }
+
+    private lookupDefinition(schemaId: string, ref: string) {
+        const schema = this.getSchema(schemaId);
+        if (!schema) {
+            return undefined;
+        }
+        return pointer.get(schema, ref);
+    }
+
+    private getDefinitionRecursive(definitionReference: string, currentSchemaId: string, visitedDefinitions: string[]) {
+        const splitReference = definitionReference.split('#');
+        let newSchemaId = splitReference[0];
+        const ref = splitReference[1];
+        visitedDefinitions.push(definitionReference);
+
+        if (!newSchemaId) {
+            newSchemaId = currentSchemaId;
+            this.logger.debug(`Resolving reference ${ref} against current schema.`);
+        }
+        this.logger.debug(`Recursively resolving the reference, ref: ${ref}`);
+        const definition = this.lookupDefinition(newSchemaId, ref);
+        if (!definition) {
+            // schema not defined
+            // TODO enforce this once we can guarantee we always have schemas available
+            return this.getMissingSchemaPlaceholder(definitionReference);
+        }
+        if (!definition['$ref']) {
+            this.logger.debug('Reached a definition with no ref, terminating recursive lookup.');
+            return definition;
+        }
+        const newRef: string = definition['$ref'];
+        if (visitedDefinitions.includes(newRef)) {
+            this.logger.warn('Circular reference detected. Terminating reference lookup. Visited definitions: ' + visitedDefinitions);
+            return definition;
+        }
+        const innerDef = this.getDefinitionRecursive(newRef, newSchemaId, visitedDefinitions);
+        return mergeSchemas(innerDef, definition);
+    }
+
+    private getMissingSchemaPlaceholder(reference: string): object {
```

Comment:
I get not wanting to write out the entire return type in instances like this where it's a bit nested and constant, so I'd suggest just taking the return type away! Typescript will at least then figure it out and not just say `object` because you overrode it to be that. It's totally valid to omit the return type in functions as trivial as this